src/engine/controls/loopingcontrol.cpp

```diff
@@ -1078,13 +1078,28 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
                 //
                 // If we press 1/2 beatloop we want loop from 50% to 100%,
                 // If I press 1/4 beatloop, we want loop from 50% to 75% etc
-                double beat_len = nextBeat - prevBeat;
-                double loops_per_beat = 1.0 / beats;
-                double beat_pos = currentSample - prevBeat;
-                int beat_frac =
-                        static_cast<int>(floor((beat_pos / beat_len) *
-                                                loops_per_beat));
-                newloopSamples.start = prevBeat + beat_len / loops_per_beat * beat_frac;
+                double beatLength = nextBeat - prevBeat;
+                double beatPos = currentSample - prevBeat;
+
+                // find the two closest beat fractions and place the new loop start to the closer one
+                double fractionBeatLength = beatLength * beats;
+                double previousFractionBeat = prevBeat + (floor(beatPos / fractionBeatLength)) * fractionBeatLength;
+                double nextFractionBeat = prevBeat + (floor(beatPos / fractionBeatLength) + 1) * fractionBeatLength;
+
+                if (abs(previousFractionBeat - currentSample) <= abs(nextFractionBeat - currentSample)) {
+                    newloopSamples.start = previousFractionBeat;
+                } else {
+                    newloopSamples.start = nextFractionBeat;
+                }
+
+                // If running reverse, move the loop one fractional beat to the left.
+                // Thus, the loop end is (time-wise) in front of the current position.
+                //
+                // TODO: There must be a better way of checking if we are running reverse. Getting the "old_reverse" from the EngineBuffer somehow seems wrong.
+                // Furthermore, it does not work when we are not playing and the reverse button is held.
+                if (getEngineBuffer()->isReverse()) {
```

Comment:
For now I have now better idea. This is the real reverse state of the last engine call that includes scratching and seeks. You actually want the future. This: ControlPushButton(ConfigKey(group, "reverse")); would be the reverse button. 
This excludes scratches and seeks, probably desired?